### ISSUE TITLE
Set and Dict generalizing the type unnecessarily

### ISSUE THREAD

**Issue by MalteEbner** [python/mypy#17238]
**Bug Report**

See these two example functions:
```python
def f(items: list[str] | list[int]) -> None:
    items_as_set: set[str] | set[int] = set(items)

def g(key_to_count: dict[str, float] | dict[int, float]) -> None:
    key_to_count_2: dict[str, float] | dict[int, float] = {k: v for k, v in key_to_count.items()}
```

Mypy 1.10.0 complains about the assignment in each of them:
```
test_mypy_dict.py:2: error: Incompatible types in assignment (expression has type "set[object]", variable has type "set[str] | set[int]")  [assignment]
test_mypy_dict.py:5: error: Incompatible types in assignment (expression has type "dict[str | int, float]", variable has type "dict[str, float] | dict[int, float]")  [assignment]
```

**To Reproduce**

Playground link: https://mypy-play.net/?mypy=latest&python=3.12&gist=83bde5f3f11470e94b6a680158649641

**Expected Behavior**

The `set()` or `dict()` operation applied on a variable that is either a `list[str] | list[int] is always either a `set[str]` or `set[int]`. The same applies to function `g` using a `dict`. Thus the functions should not raise a mypy error.

**Workarounds**

I tried two workarounds:
- Using a mixed type of list[str | int]. However, I get the type list[str] | list[int] by an external function, thus this is not compatible.
- Using a generic. This is not working either.
See the playground: https://mypy-play.net/?mypy=latest&python=3.12&gist=264490179628a0a214a2e5bd947b8ccf
![image](https://github.com/python/mypy/assets/20324507/8fed4e35-80b8-4a4e-9de7-8fe3048fcd85)
